dbt/runner.py

```diff
@@ -25,8 +25,48 @@
 
 
 class RunManager(object):
-    def __init__(self, config):
+    def __init__(self, config, query, Runner):
+        """
+        Runner is a type (not instance!) derived from
+            dbt.node_runners.BaseRunner
+        """
         self.config = config
+        self.query = query
+        self.Runner = Runner
+
+        manifest, linker = self.compile(self.config)
+        self.manifest = manifest
+        self.linker = linker
+
+        selector = dbt.graph.selector.NodeSelector(linker, manifest)
+        selected_nodes = selector.select(query)
+        self.job_queue = self.linker.as_graph_queue(manifest, selected_nodes)
+
+        # we use this a couple times. order does not matter.
+        self._flattened_nodes = [
+            self.manifest.nodes[uid] for uid in selected_nodes
+        ]
+
+        self.run_count = 0
+        self.num_nodes = len([
+            n for n in self._flattened_nodes
+            if not Runner.is_ephemeral_model(n)
+        ])
+        self.node_results = []
+        self._skipped_children = {}
```

Comment:
I think this is strictly an improvement of the RunManager. But, this branch changes it from a reusable thing to a single-use-only thing -- once you call `run()`, the queue is spent and you need a new `RunManager` to run more nodes.

Architecturally I wonder what the point of the RunManager is... the Runner knows how to execute a specific node type / command, the Queue knows how to execute a series of nodes, and the RunManager just sets up the state and knocks it down. That leads me to believe that the "RunManager" is actually an "Invocation"... it contains all the state for the invocation, and has states like "ready", "in progress", and "complete"

Reply:
I have mixed feelings about the existence of RunManager. On the one hand, something has to live "above" the individual task runners and maintain overall run state, and making that thing a class makes organization/data sharing across threads easier. On the other hand, it's hard to decide with these changes where the boundary between `__init__` and `run` ought to even be for the RunManager, and it feels like its entry point should just be a single function.

The fact that it isn't re-usable is kind of regrettable, but also it's not like we ever re-used RunManager to begin with.